### DISCUSSION TITLE
Why a middleware can not be used with a serverless function?

### DISCUSSION THREAD

**Discussion by smohadjer** [vercel/community#4521] [Feature Question]
### Question

I want to use a middleware to process requests coming to my API written using serverless functions (no framework), but I realized vercel middleware only works with edge functions which use a different runtime than Node.js. Why is that? Can't we use middlewares like those we use in express server to process requests before they reach backend Node.js functions?

**Comment by swarnava** [comment]
Hi,

A really good example of using Serverless Function Middleware can be found here:  https://github.com/vercel/next.js/tree/canary/examples/api-routes-middleware

**Reply by smohadjer** [reply]
Thanks swarnava, but where exactly is the middleware there? I only see a serverless function (api/cookies.ts) that imports another js function (utils/cookies.ts). 

Middleware is a function that is invoked before request reaching the api endpoint, not after. And I'm not using next framework.

**Comment by smohadjer** [comment]
As an example below is an Express middleware that I want to migrate to Vercel and use to authenticate requests to my serverless functions. Is it possible to use such a Node.js framework-independent middleware on Vercel with small tweaks or so? I could migrate all my API endpoints to Vercel without any changes, but I got stuck on middleware migration! :(

````
import jwt from 'jsonwebtoken';

export default (req, res, next) => {
  const authHeader = req.headers['authorization'];
  if (authHeader) {
    const token = authHeader.split(' ')[1];
    try {
      const decoded = jwt.verify(token, 'mySecret');
      next();
    } catch(err) {
      res.status(401).json({'error': 401, 'message': 'invalid token'})
    }
  } else {
    res.status(403).json({'error': 403, 'message': 'authorization header not found'});
  }
}
````
